dash/dash.py

```diff
@@ -990,15 +999,25 @@ def add_context(*args, **kwargs):
                         )
 
                     component_ids = collections.defaultdict(dict)
+                    has_update = False
                     for i, o in enumerate(output):
-                        component_ids[o.component_id][o.component_property] =\
-                            output_value[i]
+                        val = output_value[i]
+                        if val is not no_update:
+                            has_update = True
+                            o_id, o_prop = o.component_id, o.component_property
+                            component_ids[o_id][o_prop] = val
+
+                    if not has_update:
+                        raise exceptions.PreventUpdate
```

Comment:
Could use `flask.abort` instead of going the slow exception way.

Reply:
I don't know what `flask.abort` does, but I wanted to trigger the `PreventDefault` handler, in case the renderer does any optimizations with that specific response to maximize performance.

https://github.com/plotly/dash/blob/bb5ed706f4da8ff55a56dad49c90f2f07fce3651/dash/dash.py#L180-L183

Reply:
It's an empty 204 which would be the same as `flask.abort(204)`. Raising the exception takes longer because it has to be caught and then the handler is called whereas abort is immediate.

Reply:
Actually seems like `flask.abort` is doing more than that... if I try to do `flask.abort(204)` it fails with `LookupError: no exception for 204`. The [docs](http://flask.pocoo.org/docs/0.12/api/#flask.abort) say:
> If a status code is given it’s looked up in the list of exceptions and will raise that exception

Which anyway sounds at least as complicated as what we're doing with `raise PreventUpdate`.

Reply:
Right, 204 is not an error. But then you are at the dispatch level, think you can just do `return '', 204` and avoid any error handling.

Reply:
We're in the weeds now - the return from this function feeds into `response.set_data()` in the actual `dispatch`, which I don't believe accepts a status code.
https://github.com/plotly/dash/blob/b2a97b41d376baf5122e59eb7d64641608cdffa6/dash/dash.py#L1094

There may be a way to improve on this but I'd like to defer it for this PR, if that's alright.

Reply:
Ah, that's new. Anyway, this works well for partial updates.